File: C&Python/DFS/N과 M(10).py
def solution(i, cnt, temp, d):
    # exit condition
    if cnt == M and ' '.join(temp) not in d:
        d[' '.join(temp)] = 1
        print(' '.join(temp))
        return

    if i >= N:
        return

    # visit
    temp.append(str(arr[i]))
    solution(i + 1,cnt + 1, temp, d)
    temp.pop()

    # not visit
    solution(i + 1, cnt, temp, d)

# input
N, M = map(int, input().split())
arr = list(map(int,input().split()))
arr.sort()
d = {} # 결과 중복출력 방지
solution(0, 0,[], d)

File: C++/Baekjoon/Brute-force(재귀)/N과 M(10).cpp
#include <iostream> // cin, cout
#include <algorithm> //sort
#include <string> // to_string
#include <unordered_map> //stl unordered_map(hash container)
using namespace std;

int n, m;
int v[9];
int output[9];
unordered_map<string, int> um; 

void BFS(int index, int cnt, string cur_string)
{
	//exit condition
	if (cnt == m)
	{
		if (um.find(cur_string) == um.end()) // double check
		{
			for (int i = 0; i < m; i++)
			{
				cout << output[i] << " ";
			}
			cout << endl;
			um[cur_string] = 1;
		}
		return;
	}

	// next step
	for (int i = index; i < n; i++)
	{
		output[cnt] = v[i];
		BFS(i + 1, cnt + 1, cur_string + to_string(v[i]));
	}
}

int main()
{
	// input
	cin >> n >> m;
	for (int i = 0; i < n; i++)
	{
		cin >> v[i];
	}
	
	// sort
	sort(v, v + n);

	// BFS function
	BFS(0, 0,"");
	return 0;
}